kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,471 @@
+
+        Args:
+            pattern_obj (dict): The pattern object containing fields to match.
+            kcidb_obj (dict): The KCIDB object containing fields to be matched
+            against.
+
+        Returns:
+            bool: True if all fields match, False otherwise.
+        """
+
+        def match_value(pattern_value, kcidb_value):
+            """Check if a pattern value matches a KCIDB value."""
+            if isinstance(pattern_value, str) and isinstance(kcidb_value, str):
+                return bool(re.match(pattern_value, kcidb_value))
+            return pattern_value == kcidb_value
+
+        def match_item(pattern_item, kcidb_item):
+            """Match all sub-fields in pattern_item against those in
+            kcidb_item."""
+            for sub_key, sub_pattern_value in pattern_item.items():
+                if sub_key == "log_regex":  # ignore log_regex here
+                    return True
+                if sub_key not in kcidb_item:
+                    logger.debug("Field %s does not exist "
+                                 "in KCIDB entry", sub_key)
+                    return False
+                if not match_value(sub_pattern_value, kcidb_item[sub_key]):
+                    logger.debug("Field %s with value %s does not match %s",
+                                 sub_key, kcidb_item[sub_key],
+                                 sub_pattern_value)
+                    return False
+            return True
+
+        for key, pattern_value in pattern_obj.items():
+            # Ensure the key exists in the KCIDB object
+            if key not in kcidb_obj:
+                logger.debug("Key %s not found in kcidb_obj", key)
+                return False
+
+            if isinstance(pattern_value, list):
+                # Both lists are empty, consider it a match
+                if not pattern_value and not kcidb_obj[key]:
+                    continue
+                # One list is empty, consider it a mismatch
+                if not pattern_value or not kcidb_obj[key]:
+                    logger.debug("Key %s is empty in one of the objects",
+                                 key)
+                    return False
+
+                # Check if the first items in the list match
+                if not match_item(pattern_value[0], kcidb_obj[key][0]):
+                    return False
+            else:
+                # Match individual items
+                if not match_item(pattern_value, kcidb_obj[key]):
+                    return False
+
+        if 'tests' in pattern_obj and 'log_regex' in pattern_obj['tests'][0]:
+            return self.check_log_match(kcidb_obj, pattern_obj, "tests")
+        if 'builds' in pattern_obj and 'log_regex' in pattern_obj['builds'][0]:
+            return self.check_log_match(kcidb_obj, pattern_obj, "builds")
+
+        return True
+
+    def check_log_match(self, kcidb_io_object, pattern_object, key):
```

Comment:
"""When the pattern_object contains a log_regex, we first try to match it against log_excerpt, if we can't find any occurrence there, we download the entire log file and try to match on it, return True if we were able to find the log_regex, False otherwise"""